src/tools/kata-ctl/src/ops/volume_ops.rs

```diff
@@ -0,0 +1,300 @@
+        DirectVolSubcommand::Add(args) => add(&args.volume_path, &args.mount_info)?,
+        DirectVolSubcommand::Remove(args) => remove(&args.volume_path)?,
+        DirectVolSubcommand::Stats(args) => executor::block_on(stats(&args.volume_path))?,
+        DirectVolSubcommand::Resize(args) => {
+            executor::block_on(resize(&args.volume_path, args.resize_size))?
+        }
+    };
+    if let Some(cmd_result) = cmd_result {
+        println!("{:?}", cmd_result);
+    }
+
+    Ok(())
+}
+
+async fn resize(volume_path: &str, size: u64) -> Result<Option<String>> {
+    let sandbox_id = get_sandbox_id_for_volume(volume_path)?;
+    let mount_info = get_volume_mount_info(volume_path)?;
+    let resize_req = ResizeVolumeRequest {
+        size,
+        volume_guest_path: mount_info.device,
+    };
+    let encoded = serde_json::to_string(&resize_req)?;
+    let shim_client = runtimes::MgmtClient::new(sandbox_id, Some(TIMEOUT))?;
+
+    let url = runtimes::DIRECT_VOLUMN_RESIZE_URL;
+    let response = shim_client
+        .post(url, &String::from(CONTENT_TYPE_JSON), &encoded)
+        .await?;
+    let status = response.status();
+    if status != StatusCode::OK {
+        let body = format!("{:?}", response.into_body());
+        return Err(anyhow!(
+            "failed to resize volume ({:?}): {:?}",
+            status,
+            body
+        ));
+    }
+
+    Ok(None)
+}
+
+async fn stats(volume_path: &str) -> Result<Option<String>> {
+    let sandbox_id = get_sandbox_id_for_volume(volume_path)?;
+    let mount_info = get_volume_mount_info(volume_path)?;
+
+    let req_url =
+        url::form_urlencoded::Serializer::new(String::from(runtimes::DIRECT_VOLUMN_STATS_URL))
+            .append_pair(runtimes::DIRECT_VOLUMN_PATH_KEY, &mount_info.device)
+            .finish();
+
+    // sid: sandbox id
+    let shim_client = runtimes::MgmtClient::new(sandbox_id, Some(TIMEOUT))?;
+    let response = shim_client.get(&req_url).await?;
+    // turn body into string
+    let body = format!("{:?}", response.into_body());
+
+    Ok(Some(body))
+}
+
+// join_path joins user provided volumepath with kata direct-volume root path
+// the volume_path is base64-encoded and then safely joined to the end of path prefix
+fn join_path(prefix: &str, volume_path: &str) -> Result<PathBuf> {
+    if volume_path.is_empty() {
+        return Err(anyhow!("volume path must not be empty"));
+    }
+    let b64_encoded_path = base64::encode(volume_path.as_bytes());
+
+    Ok(safe_path::scoped_join(prefix, b64_encoded_path)?)
+}
+
+// add writes the mount info (json string) of a direct volume into a filesystem path known to Kata Containers.
+pub fn add(volume_path: &str, mount_info: &str) -> Result<Option<String>> {
+    let mount_info_dir_path = join_path(KATA_DIRECT_VOLUME_ROOT_PATH, volume_path)?;
+
+    // create directory if missing
+    fs::create_dir_all(&mount_info_dir_path)?;
+
+    if !mount_info_dir_path.is_dir() {
```

Comment:
Nit: is `mount_info_dir_path.is_dir()` needed? Line 108 should ensure that the `mount_info_dir_path` is a directory.